## Data Engineering

### Getting the Products

This notebook contains the strategy we sued to generate data to build the Smart Bag.

We have picked up 289 products from the Flipkart Groceries website belonging to either 'snacks and beverages' or 'packaged foods'.

The tree you can see below has been initialized to make it possible for us to access various features corresponding to a product easily. It was required for the creation of the *products* and *orders* tables.


Given below is the structure of the tree:

Major category 1

  - Minor category 1.1
    
    - Product Type 1.1.1

      - Product 1.1.1.1
      - Product 1.1.1.2
      - Product 1.1.1.3
      - Product 1.1.1.4
      - Product 1.1.1.5
    
    - Product Type 1.1.2

      - Product 1.1.2.1
      - Product 1.1.2.2
      - Product 1.1.2.3
      - Product 1.1.2.4
      - Product 1.1.2.5

      .

      .

      .
   - Product Type 1.1.n

      - Product 1.1.n.1
      - Product 1.1.n.2
      - Product 1.1.n.3
      - Product 1.1.n.4
      - Product 1.1.n.5

  - Minor category 1.2
    
    - Product Type 1.2.1

      - Product 1.2.1.1
      - Product 1.2.1.2
      - Product 1.2.1.3
      - Product 1.2.1.4
      - Product 1.2.1.5

      .

      .

      .
    
   - Product Type 1.1.n

      - Product 1.1.n.1
      - Product 1.1.n.2
      - Product 1.1.n.3
      - Product 1.1.n.4
      - Product 1.1.n.5
    
    .

    .

    .

Major category 2

  - Minor category 2.1
    
    - Product Type 2.1.1

      - Product 2.1.1.1
      - Product 2.1.1.2
      - Product 2.1.1.3
      - Product 2.1.1.4
      - Product 2.1.1.5

      .

      .

      .
    
    .

    .

    .
  
  - Minor category 2.m
    
    - Product Type 2.m.1

      - Product 2.m.1.1
      - Product 2.m.1.2
      - Product 2.m.1.3
      - Product 2.m.1.4
      - Product 2.m.1.5



Each product contains the following values in order:   

- Name
- Discounted Price
- Original MRP
- Discount Percentage
- Quantity Maginitude (in the relevant unit)
- Sponsored Flag (0 for all products in the initialization)

In [ ]:
# Defining the tree
the_tree = {
           'snacks & beverages':
            {
                'biscuits':
                {
                    'cookies':
                    [
                        ['BRITANNIA NutriChoice Oats, Orange Almond Cookies', 25, 25, 0, 75, 0],
                        ['BRITANNIA Nice Time Coconut Cookies', 23, 25, 8, 150, 0],
                        ['Dukes Kaju Kukkies Cookies', 50, 100, 50, 400, 0],
                        ['Priyagold Italiano with Crunchy Choco Chips Cookies', 65, 65, 0, 100, 0],
                        ['Priyagold Italiano with Crunchy Coconut Cookies', 65, 65, 0, 100, 0]
                    ],
                    'cream biscuits':
                    [
                        ['Cadbury Chocobakes Choc Cream Filled', 51, 60, 15, 150, 0],
                        ['Sunfeast Dark Fantasy Choco Cream Filled', 99, 140, 29, 300, 0],
                        ['OREO Choco Biscuits Cream Sandwich', 30, 30, 0, 120, 0],
                        ['BRITANNIA Milk Bikis Biscuits Cream Sandwich', 42, 45, 6, 200, 0],
                        ['OREO Cadbury Dipped Cookie Cream Sandwich', 48, 60, 20, 150, 0]
                    ],
                    'marie & digestive':
                    [
                        ['BRITANNIA Nutri Choice 5 Grain High Fibre Multigrain Biscuits Digestive', 58, 65, 10, 200, 0],
                        ['BRITANNIA Gold Marie Biscuit', 32, 32, 0, 250, 0],
                        ['Dukes Digestive', 125, 250, 50, 10000, 0],
                        ['BRITANNIA Nutri Choice High Fibre Biscuits Digestive', 153, 180, 15, 1000, 0],
                        ['Priyagold Biscuits Digestive', 35, 65, 46, 200, 0]
                    ],
                    'milk & glucose':
                    [
                        ['BRITANNIA Milk Bikis Biscuits Plain', 55, 65, 15, 500, 0],
                        ['PARLE G Gold Biscuits Plain', 112, 125, 10, 1000, 0],
                        ['PARLE G Original Gluco Biscuits Plain', 75, 75, 0, 800, 0],
                        ['PARLE Monaco Classic Plain', 102, 120, 15, 800, 0],
                        ['PARLE G Gold Plain', 250, 250, 0, 2000, 0]
                    ],
                    'salted':
                    [
                        ['BRITANNIA Little Hearts Classic Sweet & Salty', 17, 20, 15, 75, 0],
                        ['BRITANNIA 50-50 Maska Chaska Salted Biscuit', 28, 30, 6, 120, 0],
                        ['BRITANNIA Nutrichoice Sugar Free Cracker Salted Biscuit', 40, 45, 11, 300, 0],
                        ['Dukes Cheesy Nibbles Salted Biscuit', 41, 55, 25, 150, 0],
                        ['PARLE Monaco Salted Biscuit', 30, 30, 0, 200, 0]
                    ],
                    'cheeslets':
                    [
                        ['PARLE Monaco Classic Cheeselets', 60, 60, 0, 150, 0]
                    ],
                    'wafers & rusk':
                    [
                        ['BRITANNIA Toastea NA flavored Sooji Rusk', 83, 98, 15, 654, 0],
                        ['BRITANNIA Premium Sooji Rusk', 38, 40, 5, 273, 0],
                        ['Dukes Waffy Choco Wafer Rolls', 75, 150, 50, 250, 0],
                        ['PARLE Premium Elaichi Rusk', 48, 48, 0, 300, 0],
                        ['BRITANNIA Treat Wafers', 17, 25, 32, 75, 0]
                    ]
                },
                'chips, namkeen & snacks':
                {
                    'namkeen':
                    [
                        ['PARLE Aloo Bhujia', 99, 200, 50, 750, 0],
                        ['Haldiram\'s Bhujia', 99, 99, 0, 400, 0],
                        ['Haldiram\'s Navrattan', 215, 215, 0, 1000, 0],
                        ['Haldiram\'s Punjabi Tadka', 47, 48, 2, 200, 0],
                        ['KURKURE Puffcorn Yummy Cheese', 20, 20, 0, 55, 0]
                    ],
                    'chips':
                    [
                        ['Parle\'s Cream and Onion Flavour Wafers', 30, 60, 50, 150, 0],
                        ['Too Yumm! Potato Stix Aloo Chat Chips', 25, 50, 50, 75, 0],
                        ['Lay’s Magic Masala Chips', 112, 150, 25, 345, 0],
                        ['Lay\'s American Style Cream and Onion Chips', 120, 150, 20, 345, 0],
                        ['Uncle Chipps Spicy Treat Chips', 19, 20, 5, 40, 0]
                    ],
                    'popcorn, papad & fryums':
                    [
                        ['ACT II Classic Salted Popcorn', 55, 55, 0, 180, 0],
                        ['PUMPUM Panipuri Fryums 250 g', 118, 120, 1, 250, 0],
                        ['ATISH PAPAD - Pani Puri, (Ready to Fry)', 40, 50, 20, 200, 0],
                        ['ATISH PAPAD - Plain', 36, 45, 20, 150, 0],
                        ['ACT II Party Pack Butter Delite Popcorn', 120, 130, 7, 450, 0]
                    ],
                    'snacky nuts':
                    [
                        ['Flipkart Supermart Masala Peanuts', 85, 170, 50, 400, 0],
                        ['Haldiram\'s Classic Gup Shup Peanuts', 50, 50, 0, 200, 0],
                        ['Haldiram\'s Classic Nut Cracker', 48, 48, 0, 200, 0],
                        ['Haldiram\'s Masala Peanut', 48, 48, 0, 200, 0],
                        ['KALEVA Tasty', 69, 118, 41, 240, 0]
                    ]
                },
                'tea':
                {
                    'tea':
                    [
                        ['Red Label Tea Box', 130, 130, 0, 250, 0],
                        ['Taj Mahal Tea Box', 294, 305, 3, 500, 0],
                        ['Tata Teaveda Ayurvedic Goodness Brahmi, Tulsi Tea Box', 99, 126, 21, 250, 0],
                        ['Flipkart Supermart Kadak Tea Pouch', 215, 400, 46, 1000, 0],
                        ['Waghbakri Premium Leaf Tea Pouch', 140, 140, 0, 250, 0]
                    ],
                    'tea bags':
                    [
                        ['Taj Mahal Rich Spices Masala Tea Bags Box', 155, 155, 0, 25, 0],
                        ['Taj Mahal Spicy Ginger Tea Bags Box', 155, 155, 0, 25, 0],
                        ['Flipkart Supermart Green Tea Bags Box', 69, 138, 50, 25, 0],
                        ['Taj Mahal Tea Bags Box', 200, 200, 0, 100, 0],
                        ['tetley Masala Tea Bags Box', 208, 260, 20, 50, 0]
                    ],
                    'green tea':
                    [
                        ['Lipton Pure & Light Green Tea Bags Box', 540, 600, 10, 100, 0],
                        ['Lipton Honey, Lemon Green Tea Bags Box', 156, 165, 5, 25, 0],
                        ['tetley Lemon, Ginger, Mint Green Tea Bags Box', 132, 165, 20, 25, 0],
                        ['tetley Immune with Added Vitamin C Ginger, Mint, Lemon Green Tea Bags Box', 423, 550, 23, 100, 0],
                        ['Lipton Pure & Light Green Tea Bags Box', 540, 600, 10, 100, 0]
                    ],
                    'ice tea':
                    [
                        ['Nestea Peach Iced Tea Pouch', 190, 190, 0, 400, 0]
                    ]
                },
                'coffee':
                {
                    'roast & ground':
                    [
                        ['CONTINENTAL Malgudi 60 Degree Fresh Filter Coffee', 80, 90, 11, 200, 0],
                        ['CONTINENTAL Malgudi 80 Degree Fresh Filter Coffee', 94, 110, 14, 200, 0],
                        ['Deccan Gold Premium Filter Coffee', 199, 285, 30, 500, 0],
                        ['CONTINENTAL Malgudi Filter Coffee', 250, 250, 0, 500, 0],
                        ['iD Traditional Decoction Filter Coffee', 95, 95, 0, 150, 0]
                    ],
                    'instant coffee':
                    [
                        ['Sunbean Beaten Instant Coffee', 117, 130, 10, 125, 0],
                        ['Nescafe Classic Instant Coffee', 78, 78, 0, 25, 0],
                        ['CONTINENTAL Xtra Instant Coffee', 115, 115, 0, 50, 0],
                        ['BRU Instant Coffee', 170, 170, 0, 100, 0],
                        ['Tata Grand Instant Coffee', 112, 150, 25, 50, 0]
                    ]
                },
                'juices':
                {
                    'fruit drinks':
                    [
                        ['Slice Thickest Mango Drink', 60, 90, 33, 1750, 0],
                        ['Priyagold Treat Mango', 64, 65, 1, 600, 0],
                        ['Priyagold Fresh Gold Pineapple', 64, 110, 41, 1000, 0],
                        ['Paper boat Swing Slurpy Mango', 38, 40, 5, 600, 0],
                        ['Priyagold Fresh Gold Guava', 64, 110, 41, 1000, 0]
                    ],
                    'mixed fruit juices':
                    [
                        ['Real Masala Mixed Fruit', 57, 115, 50, 1, 0],
                        ['Real Fruit Juice Mixed', 97, 110, 11, 1, 0],
                        ['B Natural Mixed Fruit', 82, 110, 25, 1, 0],
                        ['Real Activ 100% Mixed Fruit Juice', 110, 135, 18, 1, 0],
                        ['Tropicana Mixed Fruit Delight', 99, 110, 10, 1, 0]
                    ],
                    'guava juices':
                    [
                        ['Real Masala Guava', 52, 105, 50, 1, 0],
                        ['Priyagold Fresh Gold Guava', 64, 110, 41, 1, 0],
                        ['Paper boat Juice - Guava', 82, 110, 25, 1, 0],
                        ['Tropicana Guava Delight Fruit Beverage', 90, 100, 10, 1, 0],
                        ['B Natural Guava', 93, 110, 15, 1, 0]
                    ],
                    'lychee juices':
                    [
                        ['Real Fruit Juice - Litchi', 104, 110, 5, 1000, 0],
                        ['B Natural Litchi', 93, 110, 15, 1000, 0],
                        ['Tropicana Litchi Delight Fruit Beverage', 20, 20, 0, 200, 0],
                        ['Paper boat Lychee Juice', 180, 180, 0, 1350, 0]
                    ],
                    'orange juices':
                    [
                        ['B Natural Orange - Juice', 77, 110, 30, 1, 0],
                        ['Real Fruit Juice - Orange', 97, 110, 11, 1, 0],
                        ['Priyagold Fresh Gold Orange', 64, 110, 41, 1, 0],
                        ['Real Activ 100% Orange Juice', 110, 135, 18, 1, 0],
                        ['Tropicana Orange Delight Fruit Beverage', 99, 110, 10, 1, 0]
                    ],
                    'apple juices':
                    [
                        ['Real Activ 100% Apple Juice', 110, 135, 18, 1000, 0],
                        ['Tropicana 100% Apple Juice', 30, 30, 0, 200, 0],
                        ['Tropicana Apple Delight Fruit Beverage', 105, 105, 0, 1000, 0],
                        ['Real Fruit Juice - Apple', 105, 105, 0, 1000, 0],
                        ['Real Fruit Juice Apple', 20, 20, 0, 180, 0]
                    ],
                    'cold pressed juices':
                    [
                        ['Real Activ 100% Apple Juice', 110, 135, 18, 1000, 0],
                        ['Tropicana 100% Apple Juice', 30, 30, 0, 200, 0],
                        ['Raw Pressery Coconut Water', 60, 60, 0, 200, 0],
                        ['Raw Pressery Sugarcane Juice', 189, 220, 14, 1000, 0],
                        ['Tropicana 100% Orange Juice', 30, 30, 0, 200, 0]
                    ],
                    'other juices':
                    [
                        ['Real Masala Mixed Fruit', 57, 115, 50, 1000, 0],
                        ['Real Masala Pomegranate', 60, 120, 50, 1000, 0],
                        ['Real Masala Guava', 52, 105, 50, 1000, 0],
                        ['Slice Thickest Mango Drink', 60, 90, 33, 1750, 0],
                        ['Priyagold Treat Mango', 64, 65, 1, 600, 0]
                    ]
                },
                'health drink mix':
                {
                    'health drink mix':
                    [
                        ['Pediasure Premium Chocolate Health Drink', 315, 315, 0, 200, 0],
                        ['Cadbury Bournvita 5 Star Magic', 325, 325, 0, 750, 0],
                        ['Cadbury Bournvita Health Drink Pouch', 299, 310, 3, 750, 0],
                        ['Cadbury Bournvita Inner strength formula', 350, 380, 7, 1000, 0],
                        ['Manna Badam Mix', 125, 250, 50, 400, 0]
                     ]
                },
                'soft drinks':
                {
                    'cold drinks':
                    [
                        ['Thums Up Can', 33, 35, 5, 300, 0],
                        ['Pepsi Black Can', 27, 30, 10, 250, 0],
                        ['MiRiNDA Plastic Bottle', 89, 90, 1, 2000, 0],
                        ['Mountain Dew Can', 27, 30, 10, 250, 0],
                        ['Pepsi Can', 27, 30, 10, 250, 0]
                    ],
                    'energy & sports drinks':
                    [
                        ['Glucose-D Vitamin D Plus Calcium Energy Drink', 149, 199, 25, 1000, 0],
                        ['GLUCON-D Instant Energy Drink', 146, 163, 10, 450, 0],
                        ['Nestle MILO Activ-Go Powder Pouch Nutrition Drink', 215, 215, 0, 400, 0],
                        ['GLUCON-D Instant Energy Drink', 310, 330, 6, 1000, 0],
                        ['GLUCON-D Energy Drink', 146, 163, 10, 450, 0]
                    ],
                    'non alcoholic drinks':
                    [
                        ['Budweiser Non-Alcoholic Can', 1, 80, 98, 330, 0],
                        ['Heineken Alcohol Free Beer Can', 52, 75, 30, 330, 0],
                        ['Heineken Non Alcohol Beer Can', 315, 450, 30, 1980, 0],
                        ['Heineken Alcohol Free Glass Bottle', 90, 120, 25, 330, 0],
                        ['Kingfisher Radler Lemon Non-Alcoholic Glass Bottle', 59, 70, 15, 300, 0]
                    ],
                },
                'instant drink mixes, squash & syrups':
                {
                    'instant drink mixes':
                    [
                        ['TANG Orange Instant Drink Mix', 114, 135, 15, 500, 0],
                        ['Nestea Iced Tea Lemon', 180, 180, 0, 400, 0],
                        ['TANG Mango Instant Drink Mix', 128, 135, 5, 500, 0],
                        ['TANG Lemon Instant Drink Mix', 128, 135, 5, 500, 0],
                        ['Rasna Instant Orange', 163, 240, 32, 750, 0]
                    ],
                    'squash':
                    [
                        ['Rasna Nagpur Orange (32 Glasses)', 44, 44, 0, 25, 0],
                        ['Rasna Alphonso Mango (32 Glasses)', 44, 44, 0, 25, 0],
                        ['Rasna 21 Orange Power Squash', 129, 140, 7, 750, 0],
                        ['Rasna Fruit Plus 21 Mango Power Squash', 129, 140, 7, 750, 0],
                        ['Rasna American Pineapple (32 Glasses)', 44, 44, 0, 25, 0]
                    ],
                    'syrups':
                    [
                        ['HERSHEY\'S Chocolate Flavor Syrup Chocolate', 83, 95, 12, 200, 0],
                        ['Rooh Afza Sharbat Rose', 148, 150, 1, 750, 0],
                        ['HERSHEY\'S Genuine Chocolate Chocolate', 193, 220, 12, 623, 0],
                        ['Veeba Fudge Topping Chocolate', 112, 161, 30, 380, 0],
                        ['LION DATES SYRUP Plain', 380, 380, 0, 1000, 0]
                    ]
                },
                'water':
                {
                    'mineral water':
                    [
                        ['Himalayan Natural Mineral Water', 13, 17, 23, 200, 0],
                        ['Himalayan Live Natural Mineral Water', 28, 35, 20, 500, 0],
                        ['Himalayan Still Mineral Water', 160, 200, 20, 750, 0],
                        ['Himalayan Still Mineral Water', 80, 100, 20, 300, 0],
                        ['Himalayan Sparkling Mineral Water', 88, 110, 20, 300, 0]
                    ]
                }
            },
           'packaged food': 
            {
                'breakfast cereals':
                {
                    'flakes':
                    [
                        ['Kellogg\'s Corn Flakes Real Almond & Honey', 525, 550, 4, 1000, 0],
                        ['Kellogg\'s Crunchy Granola Chocolate & Almonds', 330, 330, 0, 450, 0],
                        ['Kellogg\'s Chocos', 109, 125, 12, 250, 0],
                        ['Kwality Choco Flakes', 249, 499, 50, 1000, 0],
                        ['Kwality Corn Flakes', 199, 398, 50, 1000, 0]
                    ],
                    'muesli':
                    [
                        ['Kwality Crunchy Muesli Fruit N Nut', 280, 560, 50, 1000, 0],
                        ['Kellogg\'s Muesliwith 21% Fruit, Nut & Seeds', 294, 325, 9, 500, 0],
                        ['Kwality Muesli Crunchy', 280, 560, 50, 1000, 0],
                        ['Yogabar Dark Chocolate & Cranberry Muesli', 399, 499, 20, 700, 0],
                        ['Kwality Almonds, Raisins and Honey Muesli', 219, 449, 51, 700, 0]
                    ],
                    'oats':
                    [
                        ['True Elements Dark Chocolate Granola', 19, 60, 68, 60, 0],
                        ['Quaker Oats with Flavour Mix', 41, 49, 16, 200, 0],
                        ['Saffola Classic Masala Oats', 175, 195, 10, 500, 0],
                        ['Quaker Oats', 180, 195, 7, 1000, 0],
                        ['Saffola Oats, Rolled Oats, 100% natural', 89, 89, 0, 500, 0]
                    ],
                    'cereal bars':
                    [
                        ['Nature Valley Oats N Honey Crunchy Granola Bars', 306, 360, 15, 252, 0],
                        ['Nature Valley Oats and Dark Chocolate Crunchy Granola Bars', 306, 360, 15, 252, 0],
                        ['Yogabar Almond Fudge Protein Bar', 112, 125, 10, 60, 0],
                        ['RiteBite Max Protein Choco Almond Bar', 49, 65, 24, 50, 0],
                        ['RiteBite Max Protein Daily Choco Berry Bars', 273, 360, 24, 300, 0]
                    ]
                },
                'noodles & pasta':
                {
                    'noodles':
                    [
                        ['Maggi Masala Instant Noodles Vegetarian', 136, 144, 5, 84, 0],
                        ['Sunfeast Yippee Magic Masala Instant Noodles Vegetarian', 91, 144, 36, 810, 0],
                        ['Maggi 2 Minute Special Masala Instant Noodles Vegetarian', 15, 15, 0, 70, 0],
                        ['Ching\'s Secret Veg Hakka Noodles Vegetarian', 85, 107, 20, 600, 0],
                        ['Maggi Nutri licious Atta Masala Instant Noodles Vegetarian', 128, 140, 8, 435, 0]
                    ],
                    'pasta':
                    [
                        ['24 Mantra Whole Wheat Flour Pasta', 50, 80, 30, 400, 0],
                        ['24 Mantra Organic Whole Wheat Flour Pasta', 56, 80, 30, 400, 0],
                        ['Flipkart Supermart Durum Wheat Semolina Penne Pasta', 55, 160, 65, 400, 0],
                        ['Maggi Cheese Macaroni Pasta', 28, 28, 0, 70, 0],
                        ['YIPPEE Tricolor Masala Pasta', 22, 25, 12, 65, 0]
                    ],
                    'vermicelli':
                    [
                        ['MTR Vermicelli', 67, 80, 16, 850, 0],
                        ['Rajdhani Long Cut Vermicelli', 20, 20, 0, 150, 0],
                        ['Bambino Vermicelli', 82, 89, 7, 875, 0],
                        ['Rajdhani Short_Cut Vermicelli', 70, 70, 0, 900, 0],
                        ['MTR Rice Vermicelli', 50, 50, 0, 400, 0]
                    ]
                },
                'ketchups & spreads':
                {
                    'ketchups':
                    [
                        ['ChefBoss Chilli Chinese Stir Fry Sauce & Dip', 67, 75, 10, 200, 0],
                        ['ChefBoss Kung Pao Cooking Sauce', 67, 75, 10, 200, 0],
                        ['Del Monte Tomato Classic Blend Ketchup', 79, 115, 31, 800, 0],
                        ['Kissan Fresh Tomato Ketchup', 110, 120, 8, 950, 0],
                        ['SMITH & JONES Tomato Ketchup', 91, 130, 30, 1000, 0]
                    ],
                    'dips & spreads':
                    [
                        ['24 Mantra Peanut Butter Creamy', 280, 400, 30, 800, 0],
                        ['St. Dalfour Raspberry Fruit Spread', 350, 350, 0, 284, 0],
                        ['WINGREENS Premium Veg Mayo', 99, 199, 50, 800, 0],
                        ['Wingreens Farms Pizza \'N\' Pasta Sauce', 74, 149, 50, 450, 0],
                        ['VEEBA Eggless Mayonnaise Chef\'s Special', 147, 189, 22, 875, 0]
                    ],
                    'dressing sauce':
                    [
                        ['FUN FOODS American Mustard', 69, 69, 0, 260, 0],
                        ['VEEBA Chipotle Southwest Dressing Sauce', 161, 161, 0, 300, 0],
                        ['VEEBA Barbeque Sauce', 141, 141, 0, 330, 0],
                        ['VEEBA White Pasta Dressing Sauce', 99, 99, 0, 285, 0],
                        ['VEEBA Honey Mustard Dressing Sauce', 161, 161, 0, 300, 0]
                    ]
                },
                'chocolates & sweets':
                {
                    'chocolates':
                    [
                        ['BOUNTY Miniatures Bars', 179, 199, 10, 170, 0],
                        ['Cadbury Dairy Milk Silk Fruit and Nut Bars', 160, 160, 0, 137, 0],
                        ['Cadbury Dairy Milk Home Treats Chocolate Bars', 111, 149, 25, 126, 0],
                        ['Cadbury Dairy Milk Silk Oreo Chocolate Bars', 148, 175, 15, 130, 0],
                        ['Nestle Munch Bars', 100, 100, 0, 208, 0]
                    ],
                    'sweets & mithai':
                    [
                        ['Lalji Sponge Rasogolla', 157, 315, 50, 1000, 0],
                        ['BHIKHARAM CHANDMAL Rasgulla Tin - 1kg (Pack of 1) Tin', 150, 300, 50, 1000, 0],
                        ['BHIKHARAM CHANDMAL Gulab jamun Tin - (Pack of 1) Tin', 150, 300, 50, 1000, 0],
                        ['Paper boat Daily Snack Chikki Pouch', 165, 275, 40, 496, 0],
                        ['Lal Mysore Pak Classic Box', 199, 399, 50, 400, 0]
                    ],
                    'candy & gums':
                    [
                        ['Skittles Wild Berry Candy', 45, 50, 10, 33.6, 0],
                        ['Orbit Sugar Free Sweet Mint Chewing Gum', 46, 50, 8, 22, 0],
                        ['Alpenliebe Strawberry Jelly Candy', 40, 50, 20, 148, 0],
                        ['Alpenliebe Gold Rich Milky Caramel Toffee', 85, 100, 15, 390, 0],
                        ['Candyman Fruiteefun Multi Flavor', 112, 150, 25, 810, 0]
                    ],
                    'mouth freshner':
                    [
                        ['Doublemint Chewy Peppermint Mouth Freshener', 100, 100, 0, 80.85, 0],
                        ['Doublemint Chewy Lemon Mint', 100, 100, 0, 80.85, 0],
                        ['Polo Mint', 10, 10, 0, 24, 0],
                        ['Delight nuts Icecream Mukhwas Vanilla Mouth Freshener', 77, 110, 30, 150, 0],
                        ['Delight nuts Saptrangee Mukhwas Mint Mouth Freshener', 69, 99, 30, 150, 0]
                    ],
                    'cake':
                    [
                        ['Cadbury Chocobakes Chocolate Cake', 54, 60, 10, 126, 0],
                        ['BRITANNIA Gobbles Fruit Cake', 49, 55, 10, 220, 0],
                        ['BRITANNIA Treat Cocoa Creme Roll Croissants', 12, 15, 20, 45, 0],
                        ['BRITANNIA Treat Creme Roll Vanilla Croissants', 12, 15, 20, 45, 0],
                        ['winkies Swiss Roll Chocolate Cake', 68, 80, 15, 175, 0]
                    ]
                },
                'jams & honey':
                {
                    'jams':
                    [
                        ['Kissan Orange Marmalade', 165, 165, 0, 500, 0],
                        ['Kissan Mango Jam', 165, 165, 0, 490, 0],
                        ['Apis Fruit Blast', 101, 199, 49, 700, 0],
                        ['Kissan Mixed Fruit Jam', 68, 68, 0, 200, 0],
                        ['Apis Mixed Fruit Jam', 153, 170, 10, 450, 0]
                    ],
                    'honey':
                    [
                        ['Apis Honey', 247, 275, 10, 500, 0],
                        ['Dabur Honey Squeezy', 297, 330, 10, 400, 0],
                        ['Dabur Honey', 387, 430, 10, 1000, 0],
                        ['PATANJALI Honey', 165, 165, 0, 500, 0],
                        ['Apis Himalaya Honey Squeezy', 118, 120, 1, 300, 0]
                    ]
                },
                'pickles & chutney':
                {
                    'pickles':
                    [
                        ['Umadi\'s Mango Pickle', 82, 165, 50, 500, 0],
                        ['Flipkart Supermart Mixed Pickle', 90, 180, 50, 400, 0],
                        ['Flipkart Supermart Mango Pickle', 100, 200, 50, 400, 0],
                        ['Top\'s Mixed Pickle', 140, 165, 15, 950, 0],
                        ['Umadi\'s Mixed Vegetable Pickle', 82, 165, 50, 500, 0]
                    ],
                    'chutney':
                    [
                        ['Ching\'s Secret Schezwan Chutney Paste', 68, 80, 15, 250, 0],
                        ['Dabur Hommade Chatpati Chutney Paste', 55, 55, 0, 200, 0],
                        ['MOTHER\'S RECIPE Bhelpuri Chutney Paste', 50, 55, 9, 200, 0],
                        ['MOTHER\'S RECIPE Delhi Chaat Chutney Paste', 50, 55, 9, 200, 0],
                        ['Dabur Hommade Spicy Garlic Chutney Paste', 65, 65, 0, 200, 0]
                    ]
                },
                'ready to cook':
                {
                    'soups':
                    [
                        ['Knorr Tomato Chatpata Cup-a-Soup', 10, 10, 0, 14, 0],
                        ['Knorr Sweet Corn Veg Cup-a-Soup', 10, 10, 0, 10, 0],
                        ['Knorr Manchow Soup', 10, 10, 0, 12, 0],
                        ['Ching\'s Secret Instant Soup', 39, 50, 22, 76, 0],
                        ['Knorr Mixed Vegetable Cup-a-Soup', 10, 10, 0, 10, 0]
                    ],
                    'ready mixes':
                    [
                        ['AASHIRVAAD Instant Mix - Gulab Jamun', 110, 220, 50, 500, 0],
                        ['MTR Gulab Jamun Mix', 130, 130, 0, 160, 0],
                        ['AASHIRVAAD Rava Idli Instant Mix', 77, 115, 33, 500, 0],
                        ['MTR Ready Mix Rava Idli', 193, 225, 14, 1000, 0],
                        ['MTR Instant Dosa Mix', 110, 125, 12, 500, 0]
                    ],
                    'ready meals':
                    [
                        ['Fortune Bengali Bhog Khichuri', 45, 55, 18, 200, 0],
                        ['Fortune Gujarati Khichdi', 45, 55, 18, 200, 0],
                        ['MTR Vegetable Upma', 16, 25, 36, 60, 0],
                        ['Tata Q Hot and Spicy Schezwan Noodles with Veggies', 66, 99, 33, 290, 0],
                        ['MTR Poha', 16, 25, 36, 60, 0]
                    ],
                    'canned food':
                    [
                        ['Del Monte Whole Green Olives', 225, 225, 0, 450, 0],
                        ['Del Monte Baked Beans', 110, 110, 0, 450, 0],
                        ['American Garden Plain Bread Crumbs', 188, 190, 1, 425, 0]
                    ]
                },
                'cooking sauces & vinegar':
                {
                    'cooking sauce':
                    [
                        ['Ching\'s Secret Green Chilli Sauce', 68, 85, 20, 680, 0],
                        ['Ching\'s Secret Red Chilli Sauce', 68, 85, 20, 680, 0],
                        ['Ching\'s Secret Dark Soy Sauce', 49, 55, 10, 210, 0],
                        ['FUN FOODS Pasta & Pizza Sauce', 89, 89, 0, 325, 0],
                        ['WeiKFiELD Green Chilli Sauce', 49, 55, 10, 200, 0]
                    ],
                    'vinegar':
                    [
                        ['Top\'s Premium Synthetic White Vinegar', 54, 54, 0, 610, 0],
                        ['Ching\'s Secret Chilli Vinegar', 45, 50, 10, 170, 0],
                        ['DiSano Apple Cider Vinegar Filtered Vinegar', 120, 240, 50, 500, 0],
                        ['DiSano Apple Cider Vinegar With Mother Vinegar Vinegar', 169, 495, 65, 500, 0],
                        ['WeiKFiELD Chilli Vinegar', 45, 50, 10, 200, 0]
                    ]
                },
                'baking':
                {
                    'baking ingredients':
                    [
                        ['Solar Dry Active Yeast Powder', 42, 60, 30, 40, 0],
                        ['Kwality Baking Soda Powder', 20, 25, 20, 100, 0],
                        ['Solar Cocoa Powder', 50, 72, 30, 50, 0],
                        ['HERSHEY\'S Cocoa Powder', 229, 270, 15, 225, 0],
                        ['WeiKFiELD Double Action Baking Powder', 33, 33, 0, 100, 0]

                    ],
                    'ready baking mixes':
                    [ 
                        ['Betty Crocker Complete Pancake Mix', 96, 105, 8, 250, 0],
                        ['Betty Crocker Triple Choco Brownie Mix', 303, 330, 8, 425, 0],
                        ['Betty Crocker Pancake Mix', 180, 205, 10, 500, 0],
                        ['WeiKFiELD Cooker Cake Mix Chocolate', 80, 95, 15, 150, 0], 
                        ['Pillsbury Choco Cooker Cake Mix', 95, 100, 5, 159, 0]
                    ]
                }
            }
          }

In [ ]:
# Essential imports
import random
from collections import defaultdict
from tqdm import tqdm
import pandas as pd

In [ ]:
# Mounting google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Creating lists corresponding to each product feature, for creating the 'products' table
id, major_cat, minor_cat, prod_type, prod_name, disc_price, orig_price, disc_perc, size, sponsored = [], [], [], [], [], [], [], [], [], []
i = 0
for major in tqdm(the_tree.keys()):
  for minor in the_tree[major].keys():
    for ptype in the_tree[major][minor].keys():
      for name, dprice, oprice, dperc, quant, spons in the_tree[major][minor][ptype]:
        id.append(str(i))
        i += 1
        major_cat.append(major)
        minor_cat.append(minor)
        prod_type.append(ptype)
        prod_name.append(name)
        disc_price.append(dprice)
        orig_price.append(oprice)
        disc_perc.append(dperc)
        size.append(quant)
        rand = random.random()
        sponsored.append(rand <= 0.1)

len(id)

100%|██████████| 2/2 [00:00<00:00, 641.58it/s]


289

In [ ]:
print("Number of minor categories = ", len(set(minor_cat)))
print("Number of product types = ", len(set(prod_type)))

Number of minor categories =  18
Number of product types =  60


In [ ]:
# Creating the products table
products = pd.DataFrame(zip(id, major_cat, minor_cat, prod_type, prod_name, disc_price, orig_price, disc_perc, size, sponsored), columns = ['Item ID', 'Major', 'Minor', 'Type', 'Name', 'Discounted Price', 'Original price', 'Discount', 'Quantity', 'Sponsored'])
products.head()

Item ID               Major     Minor  ... Discount Quantity  Sponsored
0       0  snacks & beverages  biscuits  ...        0     75.0      False
1       1  snacks & beverages  biscuits  ...        8    150.0      False
2       2  snacks & beverages  biscuits  ...       50    400.0      False
3       3  snacks & beverages  biscuits  ...        0    100.0      False
4       4  snacks & beverages  biscuits  ...        0    100.0      False

[5 rows x 10 columns]

In [ ]:
products.shape

(289, 10)

In [ ]:
# Counting products by major category
products['Major'].value_counts()

snacks & beverages    156
packaged food         133
Name: Major, dtype: int64

In [ ]:
# Counting products by minor category
products['Minor'].value_counts()

juices                                  39
biscuits                                31
chocolates & sweets                     25
chips, namkeen & snacks                 20
breakfast cereals                       20
ready to cook                           18
tea                                     16
ketchups & spreads                      15
instant drink mixes, squash & syrups    15
noodles & pasta                         15
soft drinks                             15
baking                                  10
jams & honey                            10
coffee                                  10
pickles & chutney                       10
cooking sauces & vinegar                10
health drink mix                         5
water                                    5
Name: Minor, dtype: int64

In [ ]:
# Counting products by value of discount
products['Discount'].value_counts()

0     79
10    30
50    30
15    19
20    18
30    13
5     10
25     9
1      7
8      6
7      6
18     6
41     5
12     5
33     4
11     4
36     3
9      3
14     3
6      3
22     2
16     2
23     2
24     2
46     2
65     2
32     2
3      2
2      1
4      1
98     1
68     1
29     1
31     1
40     1
49     1
51     1
21     1
Name: Discount, dtype: int64

In [ ]:
# Counting number of sponsored (and not) products
products['Sponsored'].value_counts()

False    262
True      27
Name: Sponsored, dtype: int64

In [ ]:
products.dtypes

Item ID              object
Major                object
Minor                object
Type                 object
Name                 object
Discounted Price      int64
Original price        int64
Discount              int64
Quantity            float64
Sponsored              bool
dtype: object

In [ ]:
products.to_csv('drive/My Drive/Flipkart Grid/products.csv')

In [ ]:
products = pd.read_csv('drive/My Drive/Flipkart Grid/products.csv')

### Creating the Orders

Generating a set of orders that follow a realistic trend at an individual user level, group-level, and at the level of all users combined, was the major challenge while generating the orders.

After studying the sample data provided, we defined a set of priorities that we wanted our data to adhere to. These were:

    1) Each user should have a limited spectrum of preferences for their orders.
    2) products should be diverse enough so that most users have unique preferences, while being similar to only some others.
    3) Each product category should have a variety of products in terms of brand, quantity, discount offered, and sponsorship.

After defining these, we came up with the strategy of *controlled multi-level random selection*.

Here is what we did at each level:

- Create a new user with ID as the number of the iteration of the outermost loop. In all we have 1000 users with IDs from 0 to 999.
- Selected a number *norders* randomly from 1 to 5. This would be the number or orders for a particular user. Then, we followed the next 4 steps for each order.
- Picked up both the major categories. (We would take a subset of all the categories if there are more)
- Sampling *nmin* minor categories from all, where *nmin* is a number randomly chosen from 3 to 7. This makes our list of candidate minor categories.
- Sampling *ntype* product types from all types belonging to the candidate minor categories, where *ntype* is a number randomly chosen from 5 to 9. This makes our list of candidate product types
- Sampling *nprods* items/products from products belonging to the candidate product types, where *nprods* is a number randomly chosen from 5 to 7. This makes our final list of products for each order.


The ranges for each number selection were selected after trying multiple values, to get data that was as realistic as possible and was in line with our priorities.


In [ ]:
# Generating the orders
orders = defaultdict(list)
browsing = defaultdict(list)
for user in tqdm(range(1000)):
  norders = random.randrange(1,6,1)
  for i in range(norders):
    majors = list(the_tree.keys())
    minors, types, product_list = [], [], []
    for major in majors:
      for minor in the_tree[major].keys():
        minors.append((major, minor))
    nmin = random.randrange(3,8,1)
    minors = random.sample(minors, min(nmin, len(minors)))
    #print(nmin, minors)
    for major, minor in minors:
      for ptype in the_tree[major][minor].keys():
        types.append((major, minor, ptype))
    ntype = random.randrange(5,10,1)
    types = random.sample(types, min(ntype, len(types)))
    #print(ntype, types)
    for major, minor, ptype in types:
      for product in the_tree[major][minor][ptype]:
        product_list.append((major, minor, ptype, product))
    nprods = random.randrange(5,8,1)
    nsearches = random.randrange(4,8,1)
    #print("Number of products = ", min(nprods, len(products)))
    search_list = random.sample(product_list, min(len(product_list), nsearches))
    product_list = random.sample(product_list, min(nprods, len(product_list)))
    nsearches = random.randrange(6,10,1)
    search_list = random.sample(product_list + search_list, nsearches)
    orders[str(user)].append(product_list)
    browsing[str(user)].append(search_list)

100%|██████████| 1000/1000 [00:00<00:00, 3137.27it/s]


In [ ]:
dates = [i for i in range(1, 31)]

In [ ]:
# Here we are creating the lists corresponding to all features in the 'orders' table
# Each order is being assigned a date 
# All orders for each user are sorted by the date
users = []
items = []
ord_num = 0
order_nums = []
order_dates = []
for user in tqdm(orders):
  n_user_orders = len(orders[user])
  ord_dates = sorted(random.sample(dates, n_user_orders))
  n_ord = 0
  for order in orders[user]:
    ord_num += 1
    for item in order:
      users.append(str(user))
      items.append(str(products[products['Name'] == item[3][0]].values[0][0]))
      order_nums.append(ord_num)
      order_dates.append(ord_dates[n_ord])
    n_ord += 1

100%|██████████| 1000/1000 [00:12<00:00, 82.24it/s]


In [ ]:
# Creating the table
order_table = pd.DataFrame(zip(order_nums, users, items, order_dates), columns = ['Order ID', 'User', 'Item ID', 'Day of Month'])
order_table.head(30)

Order ID User Item ID  Day of Month
0          1    0      92             1
1          1    0     134             1
2          1    0     132             1
3          1    0      72             1
4          1    0      62             1
5          1    0      73             1
6          1    0      34             1
7          2    1       1            26
8          2    1       3            26
9          2    1      78            26
10         2    1     127            26
11         2    1     127            26
12         2    1      20            26
13         2    1       0            26
14         3    1      16            29
15         3    1       1            29
16         3    1      17            29
17         3    1     167            29
18         3    1      86            29
19         3    1      11            29
20         4    2     123             4
21         4    2      78             4
22         4    2     134             4
23         4    2     116             4
24         4    2      75             4
25         4    2      79             4
26         5    2     272             9
27         5    2     284             9
28         5    2     269             9
29         5    2     275             9

In [ ]:
order_table.shape

(18005, 4)

The number of user-product pairs = 18005

In [ ]:
order_table['Order ID'].nunique()

3008

The number of orders = 3008

In [ ]:
order_table['User'].nunique()

1000

(Sanity check) The number of users = 1000

In [ ]:
order_table['Item ID'].nunique()

279

The number orders that our technique has included = 279

This means there are 10 products that have not been ordered by anyone. This is beneficial for us, as it mimicks a real world scenario.

In [ ]:
order_table['Day of Month'].value_counts()

28    707
17    678
9     652
14    641
30    640
12    632
1     628
13    627
22    626
2     617
21    608
27    605
8     603
18    602
24    596
5     595
7     595
20    594
10    594
6     591
26    589
15    582
23    570
19    567
4     562
3     556
25    552
16    539
29    529
11    528
Name: Day of Month, dtype: int64

The number of orders per day also seems to be realistic

In [ ]:
order_table.to_csv('drive/My Drive/Flipkart Grid/orders.csv')

The following code creates an orders dictionary in which every key is an order number. Every value contains a dictionary that itself has 3 parts, namely, the user ID, the day of the month, and the list of items purchased in that order.

In [ ]:
order_table = pd.read_csv('drive/My Drive/Flipkart Grid/orders.csv').drop(columns = ['Unnamed: 0'])
order_table['Order ID'] = order_table['Order ID'].astype(str)
order_table['User'] = order_table['User'].astype(str)
order_table['Item ID'] = order_table['Item ID'].astype(str)
#print(order_table.dtypes)
order_dict = defaultdict(dict)
for ord_id, user_id, item_id, dom in order_table.values:
    if not order_dict[ord_id]:
      order_dict[ord_id]["User"] = user_id
      order_dict[ord_id]['Day_of_month'] = dom
      order_dict[ord_id]['Items'] = []
    order_dict[ord_id]['Items'].append(item_id)
#print(order_dict)

We have created a table for all users as well, storing their log-in creded

In [ ]:
users = [str(i) for i in range(1000)]
emails, passwords = [], []
for user in users:
  emails.append("user_" + user + "@example.com")
  passwords.append("password" + user)
user_table = pd.DataFrame(zip(users, emails, passwords), columns = ['User', 'Email ID', 'Password'])
user_table.head()

User            Email ID   Password
0    0  user_0@example.com  password0
1    1  user_1@example.com  password1
2    2  user_2@example.com  password2
3    3  user_3@example.com  password3
4    4  user_4@example.com  password4

In [ ]:
user_table.to_csv('drive/My Drive/Flipkart Grid/users.csv')